lib/ui/painting.dart

```diff
@@ -1606,14 +1618,150 @@ class Image extends NativeFieldWrapperClass2 {
+      );
+      _disposed = true;
+      return true;
+    }());
+    _dispose();
+
+  }
+
+  void _dispose() native 'Image_dispose';
+
+  /// Returns the native wrapper of the Image. Any calls to `native` must use
+  /// this getter to avoid passing an [_ImageHandle] to native, which will
+  /// crash.
+  Image get _unwrapped => this;
+
+  Set<_ImageHandle> _handles = <_ImageHandle>{};
+
+  /// If asserts are enabled, returns the [StackTrace]s of each open handle from
+  /// [createHandle], in creation order.
+  ///
+  /// If asserts are disabled, this method always returns null.
+  List<StackTrace>? debugGetOpenHandleStackTraces() {
+    assert(!_disposed);
+    List<StackTrace>? stacks;
+    assert(() {
+      stacks = _handles.map((_ImageHandle handle) => handle.debugStack!).toList();
+      return true;
+    }());
+    return stacks;
+  }
+
+  /// Creates a disposable handle to this image.
+  ///
+  /// The returned object behaves identically to this image, except calling
+  /// [dispose] on it will only dispose the underlying native resources if it
+  /// is the last remaining handle.
+  Image createHandle() {
+    if (_disposed) {
+      throw StateError('Object disposed');
+    }
+    final _ImageHandle handle = _ImageHandle(this);
+    _handles.add(handle);
+    return handle;
+  }
 
   @override
   String toString() => '[$width\u00D7$height]';
 }
 
+/// A disposable handle to an [Image].
+///
+/// Handles can create more handles as long as they (and the underlying image)
+/// are not disposed.
+class _ImageHandle implements Image {
+  _ImageHandle(this._image) {
+    assert(() {
+      debugStack = StackTrace.current;
+      return true;
+    }());
+  }
+
+  final Image _image;
+
+  StackTrace? debugStack;
+
+  @override
+  int get width {
+    assert(!_disposed && !_image._disposed);
+    return _image.width;
+  }
+
+  @override
+  int get height {
+    assert(!_disposed && !_image._disposed);
+    return _image.height;
+  }
+
+  @override
+  Future<ByteData?> toByteData({ImageByteFormat format = ImageByteFormat.rawRgba}) {
+    assert(!_disposed && !_image._disposed);
+    return _image.toByteData(format: format);
+  }
+
+  @override
+  Image get _unwrapped {
+    assert(!_disposed && !_image._disposed);
+    return _image;
+  }
+
+  @override
+  List<StackTrace>? debugGetOpenHandleStackTraces() {
+    assert(!_disposed && !_image._disposed);
+    return _image.debugGetOpenHandleStackTraces();
+  }
+
+
+  @override
+  Image createHandle() {
+    assert(!_disposed && !_image._disposed);
+    return _image.createHandle();
+  }
+
+  @override
+  bool _disposed = false;
+
+  @override
+  void dispose() {
+    assert(() {
+      assert(!_disposed && !_image._disposed);
+      assert(_image._handles.contains(this));
+      _disposed = true;
+      return true;
+    }());
+    _image._handles.remove(this);
+    if (_image._handles.isEmpty) {
+      _image.dispose();
+    }
```

Comment:
this means that if you get a dart:ui.Image, you _must_ get a handle to it, you can just vent handles to it, because if you give someone a handle and they dispose it, your object is killed. That seems weird. Would it make sense to always vend an ImageHandle, and just make the current dart:ui.Image private to dart:ui?

This would have another advantage, which is that this patch has a hidden cost; we're making all uses of dart:ui.Image use a more expensive polymorphic dispatch whereas before they were single-dispatch (which is much cheaper). By having two unrelated classes, we would avoid this.

Reply:
Yeah, I suspect having both `Image` and `ImageHandle` public will create confusion.

Reply:
@goderbauer and I were talking about this and were a little concerned about what this does in `FrameInfo.image`, since we'll have to cache the `_ImageHandle` in there.

It makes the ownership model _slightly_ more confusing, since `FrameInfo` doesn't get disposed. The real question becomes "do I have to call createHandle on the result of `FrameInfo.image` before using it, and who has to dispose the result of it?".

Reply:
And in case it's not clear, if we leave `FrameInfo.image` alone, it all sort of works. When you get the image there, you immediately create a handle from it, and whoever else needs one does the same, and when all of you are done the underlying image gets disposed and you're not left wondering whether you're supposed to dispose the property of a class that was given to you.

Reply:
I would have preferred to only expose Handles to the framework, but the ownership of the very first handle (provided in the FrameInfo) is kinda strange. Who is in charge of disposing that one? And when would that happen?

Reply:
I've pushed up some changes that I think make sense, although I'm now realizing I forgot to update documents.

I think the way to reason about this should be if you can access an `Image`, you should assume it's yours and you _must_ dispose it.

Put another way, if you want to give an image reference to someone, you need to call `createHandle` on your handle, you should not expect them to. You can then also decide whether you want to "move" the reference to them (give it without calling `createHandle`, and now they must call dispose and you must not) or  "copy" it (call createHandle, you each dispose your instance).